cpp/src/arrow/compute/kernels/scalar_cast_test.cc

```diff
@@ -2217,6 +2217,78 @@ TEST(Cast, ListToListOptionsPassthru) {
   }
 }
 
+static void CheckStructToStruct(
+    const std::vector<std::shared_ptr<DataType>>& value_types) {
+  for (const auto& src_value_type : value_types) {
+    for (const auto& dest_value_type : value_types) {
+      std::vector<std::string> field_names = {"a"};
+      std::shared_ptr<Array> a1, b1, a2, b2;
+      a1 = ArrayFromJSON(src_value_type, "[1, 2]");
+      a2 = ArrayFromJSON(dest_value_type, "[1, 2]");
```

Comment:
We should make these arrays longer so that CheckCast/CheckScalar will actually test sliced inputs.